File: src/main/java/org/intellij/sdk/language/jass/JassReferenceContributor.java
package org.intellij.sdk.language.jass;

import com.intellij.openapi.util.TextRange;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.*;
import com.intellij.util.ProcessingContext;
import org.jetbrains.annotations.NotNull;

import static org.intellij.sdk.language.jass.JassAnnotator.SIMPLE_PREFIX_STR;
import static org.intellij.sdk.language.jass.JassAnnotator.SIMPLE_SEPARATOR_STR;

final class JassReferenceContributor extends PsiReferenceContributor {
    @Override
    public void registerReferenceProviders(@NotNull PsiReferenceRegistrar registrar) {
        registrar.registerReferenceProvider(PlatformPatterns.psiElement(PsiLiteralExpression.class),
                new PsiReferenceProvider() {
                    @Override
                    public PsiReference @NotNull [] getReferencesByElement(@NotNull PsiElement element,
                                                                           @NotNull ProcessingContext context) {
                        PsiLiteralExpression literalExpression = (PsiLiteralExpression) element;
                        String value = literalExpression.getValue() instanceof String
                                ? (String) literalExpression.getValue()
                                : null;

                        if ((value != null && value.startsWith(SIMPLE_PREFIX_STR + SIMPLE_SEPARATOR_STR))) {
                            TextRange property = new TextRange(SIMPLE_PREFIX_STR.length() + SIMPLE_SEPARATOR_STR.length() + 1, value.length() + 1);
                            return new PsiReference[]{new JassReference(element, property)};
                        }
                        return PsiReference.EMPTY_ARRAY;
                    }
                });
    }
}

File: src/main/kotlin/raft/war/language/jass/JassReferenceContributor.kt
package raft.war.language.jass

import com.intellij.openapi.util.TextRange
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.*
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import com.intellij.util.ProcessingContext
import com.intellij.util.containers.OrderedSet
import raft.war.language.jass.psi.*
import raft.war.language.jass.psi.funName.FUN_NAME_KEY

internal class JassReferenceContributor : PsiReferenceContributor() {

    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {

        registrar.registerReferenceProvider(PlatformPatterns.psiElement(JassTypes.STR),
            object : PsiReferenceProvider() {
                override fun getReferencesByElement(
                    element: PsiElement,
                    context: ProcessingContext
                ): Array<PsiReference> {

                    val myText = executeFuncName(element)
                    if (myText == null) return PsiReference.EMPTY_ARRAY

                    val result = OrderedSet<PsiElement>()
                    
                    val ref = object : JassReferenceBase(element, TextRange(1, element.textLength - 1)) {
                        override fun handleElementRename(newElementName: String): PsiElement {
                            val strval = JassElementTextFactory.getStrVal(element.project, newElementName)
                            element.firstChild.replace(strval)
                            return element
                        }

                        override fun isReferenceTo(element: PsiElement): Boolean = element.text == myText
                        override fun resolveDeclaration(): List<PsiElement> {
                            val scope = GlobalSearchScope.allScope(element.project)
                            StubIndex.getElements(
                                FUN_NAME_KEY,
                                myText,
                                element.project,
                                scope,
                                JassNamedElement::class.java,
                            ).forEach {
                                when (it.parent) {
                                    is JassFunHead, is JassNativ -> result.add(it)
                                }
                            }
                            return result
                        }
                    }

                    return arrayOf(ref)
                }
            })
    }
}

fun executeFuncName(element: PsiElement): String? {
    if (element.text.length < 3) return null
    val myText = element.text.substring(1, element.text.length - 1)

    val arglist = element.parent.parent
    if (arglist !is JassArgList) return null
    val call = arglist.parent
    if (call !is JassFunCall) return null
    val callName = call.funName.text
    if (callName != "ExecuteFunc" && callName != "ExecuteFuncEx") return null

    return myText
}
